### PR TITLE
Added ability to parse new lassen rewrite rule format

### PR DESCRIPTION
**Pull request by jack-melchert**
Also added custom nodes for all floating point and transcendental operations

### CODE REVIEW

File: metamapper/coreir_mapper.py [hunk 1]
```diff
@@ -77,49 +50,59 @@ def gen_rules(self, ops, rule_file=None, rrules=None):
...
-                        if counter_example is not None:
-                            print(counter_example)
-                            raise ValueError(f"RR for {op} fails with ^ Counter Example")
-                        self.table.add_peak_rule(new_rewrite_rule)
         else:
             for ind, peak_rule in enumerate(rrules):
-                self.table.add_peak_rule(peak_rule, name="test_name_" + str(ind))
+                if ops != None:
+                    op = ops[ind]
+                    print(op)
```
Comment (rdaly525) [inline metamapper/coreir_mapper.py]: remove print

File: metamapper/coreir_mapper.py [hunk 2]
```diff
@@ -77,49 +50,59 @@ def gen_rules(self, ops, rule_file=None, rrules=None):
...
+                    self.table.add_peak_rule(self.CoreIRNodes, peak_rule, None)
+            self.table.sort_rules()
 
-    def do_mapping(self, dag, convert_unbound=True, prove_mapping=True) -> coreir.Module:
+    def do_mapping(self, dag, kname="", convert_unbound=True, prove_mapping=True, node_cycles=None) -> coreir.Module:
         #Preprocess isolates coreir primitive modules
         #inline inlines them back in
-        #print("premapped")
-        #print_dag(dag)
+        # print("premapped")
```
Comment (rdaly525) [inline metamapper/coreir_mapper.py]: Remove commented prints in this block

File: metamapper/coreir_util.py [hunk 3]
```diff
@@ -251,41 +290,52 @@ def type_recurse(w):
...
+                    #TODO HACK assuming 16 bit constants always
+                    return ht.BitVector[16]
                 else:
-                    node = node_t(*children, iname=iname, type=sink_adt)
-                return node
+                    raise NotImplementedError()
+
+
+
+            #elif inst.module.name != "rom2" and inst.module.name != "Mem":
```
Comment (rdaly525) [inline metamapper/coreir_util.py]: Remove commented custom rom code

File: metamapper/coreir_util.py [hunk 4]
```diff
@@ -474,44 +578,75 @@ def visit_Combine(self, node: Combine):
...
+        if isinstance(node, RegisterSink):
+            cmod_t = self.get_coreir_reg(node.type)
+            config = CoreIRContext().new_values(fields={"init":node.type(0)})
             inst = self.def_.add_module_instance(node.iname, cmod_t, config=config)
         else:
+            cmod_t = self.nodes.coreir_modules[type(node).node_name]
             inst = self.def_.add_module_instance(node.iname, cmod_t)
         return inst
 
+        # create new instance
```
Comment (rdaly525) [inline metamapper/coreir_util.py]: Remove

File: metamapper/delay_matching.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    def visit_PipelineRegister(self, node):
+        Visitor.generic_visit(self, node)
+        child = list(node.children())[0]
+        if self.aggregate_latencies[child] is None:
+            raise ValueError("Child of pipe register is constant")
+        self.aggregate_latencies[node] = self.aggregate_latencies[child] + 1
+
+
+node_latencies = {}
```
Comment (rdaly525) [inline metamapper/delay_matching.py]: Why are there baked in latencies? This should be loaded per specific chip or be called a chip-specific Visitor Pass

File: metamapper/io_tiles.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+
```
Comment (rdaly525) [inline metamapper/io_tiles.py:1]: This looks specific to Lassen. Can this be moved?

File: metamapper/irs/coreir/ir.py [hunk 7]
```diff
@@ -2,10 +2,1031 @@
 from hwtypes import BitVector, Bit
 from hwtypes.adt import Product
 from peak import Peak, name_outputs, family_closure, Const
-from peak.family import AbstractFamily
+from peak.family import AbstractFamily, MagmaFamily, SMTFamily
+from ...node import Nodes, Constant, DagNode, Select
+from hwtypes import SMTFPVector, FPVector, RoundingMode
+import magma
 
 def gen_peak_CoreIR(width):
```
Comment (rdaly525) [inline metamapper/irs/coreir/ir.py:11]: Can you split this file? Specifically group all the adhoc micro ops for floating point in a separate included file.

**Review by rdaly525 — CHANGES_REQUESTED** [review]

File: metamapper/delay_matching.py (hunk 5 above)
Comment (jack-melchert) [reply metamapper/delay_matching.py]: We can actually remove this code as it was for a previous experiment and isn't currently being used

**Review by rdaly525 — APPROVED** [review]

**Pull request merged**